kotlinpoet/src/main/java/com/squareup/kotlinpoet/CodeWriter.kt

```diff
@@ -566,6 +569,30 @@ internal class CodeWriter constructor(
     }
   }
 
+  /**
+   * Returns whether a [KModifier.PUBLIC] should be emitted.
+   *
+   * This will return true when [KModifier.PUBLIC] is one of the [implicitModifiers] and
+   * there are no other opposing modifiers (like [KModifier.PROTECTED] etc.) supplied by the
+   * consumer.
+   */
+  private fun shouldEmitPublicModifier(
```

Comment:
Would it be possible to instead remove `PUBLIC` from every kind's implicit modifiers? That would hopefully eliminate the need for this code.

Reply:
Yeah that was another approach I considered. We'd still have to do the check for clashing modifiers(if user entered PRIVATE as one of the modifiers) so this function would still exist, just with two less if statements. 

If that's fine, I can update the PR

Reply:
I realized there are actually two ways we can approach this:

- Emit `public` if there are no clashing modifiers
- Require `KModifier.PUBLIC` to be passed in explicitly for it to be emitted

I'd actually prefer the latter. Our current logic is that we will never emit `public`, even when it's passed in explicitly - since it's an implicit modifier. I'd want us to instead honor any visibility modifier if present, and we won't emit one if none are provided. What do you think?